osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/threads/RecoverKeyProcess.java

```diff
@@ -129,65 +95,24 @@ private void checkState() {
     }
   }
 
-  private boolean canConnectUsingNewKeys() {
+  private boolean canConnectUsingNewKeys(final DlmsDevice device) {
     DlmsConnection connection = null;
     try {
-      connection = this.createConnectionUsingNewKeys();
-      return true;
+      connection =
+          this.hls5Connector.connectUnchecked(
+              device, null, this.secretManagementService::getNewKey);
```

Comment:
In the flows without key recovery it looks like connections are made via the `DlmsConnectionMessageProcessor` where `createConnectionForDevice(DlmsDevice, MessageMetadata)` surrounds the creation of the connection with calls to `throttlingService.openConnection()` and `throttlingService.closeConnection()`.

Within such a flow the `Hls5Connector` could initiate key recovery by calling `recoverKeyProcessInitiator.initiate(deviceIdentification, ipAddress)` which schedules the key recovery for asynchronous execution. If I am not mistaken, that would mean that the connection is very likely closed as far as throttling is concerned, when the recovery process comes into action.

Within the `RecoverKeyProcess` it looks like a connection is created like in the commented line without registering this activity with the throttling service. Could this lead to creating a greater number of connections at a time than allowed by the throttling budget?

Reply:
Added calls to the ThrottlingService in RecoveryKeyProcess